### ISSUE TITLE
Rails Test log says my templates were rendered when they actually weren't

### ISSUE THREAD

**Issue by tenderlove** [rspec/rspec-rails#1332]
When I run a controller spec, the test log says that my templates were rendered when they actually weren't rendered, also the response body is an empty string.

Steps to reproduce:

```
$ rails new myapp
$ cd myapp
$ echo 'gem "rspec-rails", :group => [:development, :test]' >> Gemfile
$ bundle
$ rails generate scaffold Widgets name:string
$ rake db:migrate && rake db:test:prepare
$ rake spec
$ rm log/test.log
$ bundle exec rspec spec/controllers/widgets_controller_spec.rb
```

After the final run, cat `log/test.log`.  Many times the log file will say something like:

```
Rendered widgets/new.html.erb
```

When in fact my templates were not rendered at all.  Since the templates were not actually rendered, these logs should not be in the file.

Secondly, calling `response.body` in my spec always returns a blank string.  If the templates were not rendered, and there isn't actually a response body, I'd expect either a message in the response body telling me that _or_ the method to not be available on the response. (side note: if controller tests are meant to just test the controller, why is the response object available at all?)

I realize now that I'm supposed to add `render_views` to my test to get it to render views, but as someone new to RSpec, the behavior _without_ `render_views` in my spec is extremely confusing.

**Comment by searls** [comment]
:+1:

**Comment by sgrif** [comment]
Sounds like a Rails bug. Someone should inform the core team. :trollface:

**Comment by cupakromer** [comment]
Sorry this was a confusing experience. :frowning: 

> I'd expect either a message in the response body telling me that or the method to not be available on the response.

I think a message in the body is probably the way to go. I don't feel comfortable removing / redefining a method on `ActionController::TestResponse`.

> I realize now that I'm supposed to add `render_views` to my test to get it to render views, but as someone new to RSpec, the behavior without `render_views` in my spec is extremely confusing.

I'd like to understand more why is this confusing behavior. Is this something we need to improve in the documentation? Or does the confusion stem from your long use of Minitest, where your expectations of how `ActionController::TestCase` works is conflicting with our intention of a "controller" spec?

While controller specs _are_ a thin shim on top of `ActionController::TestCase`. Some functionality, as in `render_views`, isn't run by default. The intent is treat the controller specs more in an "isolated" manner, only encouraging testing the controller's contract. We offer `view` specs and `request` specs for isolated view and fuller integration testing respectively.

**Comment by cupakromer** [comment]
> side note: if controller tests are meant to just test the controller, why is the response object available at all?

We shim on top of `ActionController::TestCase`. Rails is fairly complex. This stems from both legacy design and the desire to not re-write large sections of Rails which need to be maintained with every version update. From my current understanding the life cycle for these types of tests results in a returned response object.

I do like your idea of _not_ having the `response` object available. However, I think this isn't feasible for legacy reason (at least not until an rspec-rails 4.x), having to further customize the rails test case base setup, and would require a large shift for all existing projects.

However, I would like to explore the idea of adding accessors, even if they just ask the response object, for common key things people care about in controller specs.

**Comment by tenderlove** [comment]
> I'd like to understand more why is this confusing behavior. Is this something we need to improve in the documentation? Or does the confusion stem from your long use of Minitest, where your expectations of how ActionController::TestCase works is conflicting with our intention of a "controller" spec?

My confusion is very simple.  I did not know about the method before.  When I read the logs, it said my template was rendered when in fact it was not.  Since I thought the template was rendered, I did not think the response body would be empty.  The signs pointed to "your template was rendered successfully", not "hey you should read the docs".  Now that I know about this feature, it won't be confusing anymore.  But for someone stepping in to a legacy code base, and adding regression tests, it was very surprising.

I think just getting rid of the logs would be a huge help for someone like me.  If the "rendered template" logs weren't there, I wouldn't be surprised if the body is empty.  Even if we couldn't *remove* the lines, maybe we could have RSpec-Rails log something to the effect that it intercepted that template.  WDYT?

I think removing the body method or response object would be the most invasive / difficult way (though probably more complete than logging).

**Comment by cupakromer** [comment]
> My confusion is very simple. I did not know about the method before. When I read the logs, it said my template was rendered when in fact it was not.

Thank you :heart: for that clarification. This makes a lot of sense.

> I think just getting rid of the logs would be a huge help for someone like me. If the "rendered template" logs weren't there, I wouldn't be surprised if the body is empty. Even if we couldn't remove the lines, maybe we could have RSpec-Rails log something to the effect that it intercepted that template. WDYT?

Yes I agree. :+1:

**Comment by mrageh** [comment]
@samphippen I think this issue has been resolved by #1732

**Comment by samphippen** [comment]
@mrageh correct! Thanks ^_^
